contracts/oracles/CurveOracle.sol

```diff
@@ -0,0 +1,60 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.11;
+pragma abicoder v1;
+
+import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "../interfaces/IOracle.sol";
+import "../libraries/Sqrt.sol";
+
+import "../interfaces/ICurveRegistry.sol";
+import "../interfaces/ICurveProvider.sol";
+import "../interfaces/ICurveSwap.sol";
+
+contract CurveOracle is IOracle {
+    using Sqrt for uint256;
+
+    ICurveProvider public immutable addressProvider;
+    IERC20 private constant _NONE = IERC20(0xFFfFfFffFFfffFFfFFfFFFFFffFFFffffFfFFFfF);
+
+    constructor(ICurveProvider _addressProvider) {
+        addressProvider = _addressProvider;
+    }
+
+    function getRate(IERC20 _srcToken, IERC20 _dstToken, IERC20 connector) external view override returns (uint256 rate, uint256 weight) {
+        require(connector == _NONE, "connector should be NONE");
+
+        address srcToken = address(_srcToken);
+        address dstToken = address(_dstToken);
+        uint256 index = 0;
+        ICurveRegistry registry = ICurveRegistry(addressProvider.get_address(0));
+        address pool = registry.find_pool_for_coins(srcToken, dstToken, index);
+
+        require(pool != address(0), "CO: no pools");
+
+        uint256 b0;
+        uint256 b1;
+
+        while (pool != address(0)) {
+            (int128 srcTokenIndex, int128 dstTokenIndex, bool isUnderlying) = registry.get_coin_indices(pool, srcToken, dstToken);
+            b0 = 10 ** ERC20(srcToken).decimals();
+            if (!isUnderlying) {
+                b1 = ICurveSwap(pool).get_dy(srcTokenIndex, dstTokenIndex, b0);
+            } else {
+                b1 = ICurveSwap(pool).get_dy_underlying(srcTokenIndex, dstTokenIndex, b0);
+            }
+
+            uint256 w = b0 * b1;
```

Comment:
For weight it'll be better to use pool balances. Like this:

```solidity
balances = registry.get_balances(pool);
w = balances[srcTokenIndex] * balances[dstTokenIndex];
```

Or in case of underlying:
```solidity
balances = registry.get_underlying_balances(pool);
w = balances[srcTokenIndex] * balances[dstTokenIndex];
```

Reply:
Though at the moment there is probably a bug in `registry.get_underlying_balances` implementation, so to get actual balance you'll additionally need to scale `underlying_balance` by `10 ** (18 - token.decimals())` like this:

```solidity
balances = registry.get_underlying_balances(pool);
w = balances[srcTokenIndex] / (10 ** (18 - srcToken.decimals())) * balances[dstTokenIndex] / (10 ** (18 - dstToken.decimals()));
```

Reply:
is this a necessary change for it to work? this will look pretty ugly since the token indexes are `int128` and they first need to be converted to `uint256`. since explicit/implicit conversion is not allowed some workaround will have to be used. thoughts?

Reply:
@coolhill check my PR https://github.com/coolhill/offchain-oracle/pull/2 I fixed the weight issue and added support for newer curve pools as they have incompatible ABIs. Also changed tokens a bit in tests to speed them up.

Reply:
LGTM!